### DISCUSSION TITLE
Running OpenVSLAM on Raspberry Pi

### DISCUSSION THREAD

**Discussion by jaytxrx** [stella-cv/stella_vslam#65] [Q&A] [answered]
Hello all,

is it possible to run OpenVSLAM on Raspberry Pi 4 ? I see some demo for Jetson Nano in internet. But never seen for Raspberry Pi.

Has anyone tried it ? Is Raspberry Pi less powerful for OpenVslam ?

Thanks

**Comment by ThomasHauth** [comment, marked as answer]
Hey @jeffreyantony ,
I have been using OpenVSLAM successfully on a Raspberry Pi4. However, there are obviously a couple of limitations on such a platform if you want to run OpenVSLAM in real-time. In my experience, to achieve real-time processing, you need to use a framerate of around 10 fps and only mono-camera images (stereo image processing is too demanding) with a resolution of around 640x480. Also you need to lower the amount of tracked feature points to below 1000.

Hope that helps you,
Thomas

**Reply by jaytxrx** [reply]
@ThomasHauth 
I haven't used OpenVSLAM. But it sounds to me that 10fps and 1000 tracked feature points are a configuration parameter in OpenVSLAM. Is this correct ?

Also how do you send the raspberry pi images to OpenVSLAM ? Are direct camera inputs supported or is it just done by sending one camera image at a time or as a video file ?

How to set the camera resolution to 640x480 ? Is it an option in OpenVSLAM ?

**Reply by surfii3z** [reply]
Usually, the FPS depends on the rate that your camera streams the image frame.  But, you also need to specify it in the OpenVSLAM configuration, for example.
https://github.com/OpenVSLAM-Community/openvslam/blob/00cdcc6d5c979c0ebd461f59974b4aa43e3c1a2b/example/kitti/KITTI_mono_00-02.yaml#L22

The number of track features is the parameter, which you can set by yourself, of OpenVSLAM, for example

https://github.com/OpenVSLAM-Community/openvslam/blob/00cdcc6d5c979c0ebd461f59974b4aa43e3c1a2b/example/kitti/KITTI_mono_00-02.yaml#L32

If you want to run the SLAM `live`, you have two choices.

1) Using `ROS`, you can send the each frames of the image to the OpenVSLAM node using topics. See [this](https://openvslam-community.readthedocs.io/en/latest/ros.html) for the tutorial 

2) Without ROS, you need to send each image frames from your raspberry pi to OpenVSLAM.
See [this source code](https://github.com/OpenVSLAM-Community/openvslam/blob/00cdcc6d5c979c0ebd461f59974b4aa43e3c1a2b/example/run_camera_slam.cc) and [this turorial](https://openvslam-community.readthedocs.io/en/latest/example.html#id5) for example.

Someone please correct me if i'm wrong.

**Reply by ThomasHauth** [reply]
@surfii3z thanks for the detailed explanation. Let me just add that using the ROS video streaming might not work well on the Pi because the overhead of this method is quite large.
In my case I was implementing a custom readout using the raspberry pi's hardware camera interface (MIPI) and also dropped some frames to arrive at an input frame rate of 10 for openvslam.
